blockdevice/encryption/luks/luks.go

```diff
@@ -0,0 +1,238 @@
+	default:
+		return 0, fmt.Errorf("unknown cipher kind %s", s)
+	}
+}
+
+const (
+	// AESXTSPlain64CipherString string representation of aes-xts-plain64 cipher.
+	AESXTSPlain64CipherString = "aes-xts-plain64"
+	// AESXTSPlain64Cipher represents aes-xts-plain64 encryption cipher.
+	AESXTSPlain64Cipher Cipher = iota
+)
+
+// LUKS implements LUKS2 encryption provider.
+type LUKS struct {
+	cipher                 Cipher
+	argonBenchmarkDeadline time.Duration
+	argonIterations        uint
+	argonMemoryLimit       uint64
+}
+
+// New creates new LUKS2 encryption provider.
+func New(cipher Cipher, argonBenchmarkDeadline time.Duration, argonIterations uint, argonMemoryLimit uint64) *LUKS {
+	return &LUKS{
+		cipher:                 cipher,
+		argonBenchmarkDeadline: argonBenchmarkDeadline,
+		argonIterations:        argonIterations,
+		argonMemoryLimit:       argonMemoryLimit,
+	}
+}
+
+// Open runs luksOpen on a device and returns mapped device path.
+func (l *LUKS) Open(deviceName string, password []byte) (string, error) {
+	parts := strings.Split(deviceName, "/")
+	mappedPath := util.PartPathEncrypted(parts[len(parts)-1])
+	parts = strings.Split(mappedPath, "/")
+	mappedName := parts[len(parts)-1]
+
+	err := l.runCommand([]string{"luksOpen", deviceName, mappedName, "--key-file=-"}, password)
+	if err != nil {
+		return "", err
+	}
+
+	return mappedPath, nil
+}
+
+// Encrypt implements encryption.Provider.
+func (l *LUKS) Encrypt(deviceName string, passwords ...interface{}) error {
+	if len(passwords) == 0 || len(passwords)%2 != 0 {
+		return fmt.Errorf("expected slot:password pairs list")
+	}
+
+	cipher, err := l.cipher.String()
+	if err != nil {
+		return err
+	}
+
+	getPassword := func(index int) (int, []byte, error) {
+		s, ok := passwords[index].(int)
+		if !ok {
+			return 0, nil, fmt.Errorf("failed to parse the first key slot number")
+		}
+
+		p, ok := passwords[index+1].([]byte)
+		if !ok {
+			return 0, nil, fmt.Errorf("failed to parse the first passphrase")
+		}
+
+		return s, p, nil
+	}
+
+	slot, password, err := getPassword(0)
+	if err != nil {
+		return err
+	}
+
+	args := []string{"luksFormat", "--type", "luks2", "--key-file=-", "-c", cipher, fmt.Sprintf("--key-slot=%d", slot), deviceName}
+
+	if len(passwords) == 0 {
+		return fmt.Errorf("luksFormat expects at least one encryption password")
+	}
+
+	err = l.runCommand(args, password)
+	if err != nil {
+		return err
+	}
+
+	for i := 2; i < len(passwords); i += 2 {
+		var extraPassword []byte
+
+		slot, extraPassword, err = getPassword(i)
+		if err != nil {
+			return err
+		}
+
+		if err = l.AddKey(deviceName, slot, password, extraPassword); err != nil {
+			return err
+		}
+	}
+
+	return err
+}
+
+// Close implements encryption.Provider.
+func (l *LUKS) Close(devname string) error {
+	return l.runCommand([]string{"luksClose", devname}, nil)
+}
+
+// AddKey adds a new key at the LUKS encryption slot.
+func (l *LUKS) AddKey(devname string, slot int, password, newPassword []byte) error {
```

Comment:
is it `newSlot`? or "old" for the `password`?

Reply:
`newSlot`, yeah. Old password is searched in all keyslots.